### ISSUE TITLE
RxSwift: requestWithProgress().filterCompleted() no onNext events

### ISSUE THREAD

**Issue by lordcodes** [Moya/Moya#1771]
Moya version: 11.0.2, but also tested with Moya 12

I am using Moya with RxSwift, using the `requestWithProgress` function to get the response and progress on uploading a file.

The progress part is working perfectly, however, I found that when I used the response, my other Rx operations on the provided Observable weren't working correctly. When I debugged what was going on, I noticed that the Observable I received from doing:

```swift
requestWithProgress(...).filterCompleted()
```

was only firing an `onCompleted` and then dispose. There was no `onNext` that contained the actual response.

To get this working instead of using the provided `filterCompleted` I just had to do it myself with:

```swift
requestWithProgress(...)
  .map { $0.response }
  .filter { $0 != nil }
  .map { $0! }
```

In short, this is pulling out the response field, filtering for when it isn't nil and then unwrapping the optional.

When I use this instead, everything works as expected. I think the `filterCompleted` implementation is firing no `onNext` events. The type `Observable<Response>` and the comment on the function suggest that it should be providing an Observable that gives you the response and then an onCompleted event, so I think there is a bug there.

For now my code is working with what I put above, so its not blocking me from using this, but just raising this issue, as I think there is a bug that may be affecting others.

**Comment by lordcodes** [comment]
Thanks, will have a look when I get a chance.

I'm just using my own version right now rather than using filterCompleted,
so at least have an alternative for the time being.

Appreciate the links!

**Comment by ryancrunchi** [comment]
I've also run into this issue today. I have logged the `progressObject` and found this :
`Fraction completed: 0.0000 / Completed: 243 of -1`
It seems that the progress's `totalUnitCount` remains `-1` while `completedUnitCount` increases. Which then leads to a `fractionCompleted` being always `0`.

- `progress` property of ProgressResponse is never `1` (based on progressObject's `fractionCompleted`)
- `completed` property is never `true` (based on `progress` value)
- `filterCompleted()` never succeeds (based on `completed`)

The issue surely comes from the progressObject totalUnitCount being -1.
My thought is that with small responses, the request is completed in one batch (one delegate call). The progress `totalUnitCount` is initialized at -1 and is not updated with the response size before the request completes.
This is a theory I did not have time to check further this issue.

**Comment by AndrewSB** [comment]
That is super interesting, thanks for digging @ryancrunchi. How do you think we should handle the issue?

If we don't know what the total payload size is, I can't think of any better value than 0 to call with for progress. Would you agree?

**Comment by ryancrunchi** [comment]
Yes, I agree, progress should remain 0.
But the `filterCompleted` should rely on an other condition than the `fractionCompleted` value.

**Comment by AndrewSB** [comment]
yeah, that makes sense. If anyone would like to take a shot at fixing the behavior, I'd be more than happy to review a PR

**Comment by ryancrunchi** [comment]
I don't know what the state of this issue now, but I have tested that when a download is not complete, `progressResponse.response` is `nil`. Once the download is complete, `progressResponse.response` is not nil.
Maybe we could rely on this to filter the completed state of the download.

**Comment by sunshinejr** [comment]
@ryancrunchi @AndrewSB FYI I've created a PR with a fix (#1815).

**Comment by sunshinejr** [comment]
Closing it as the PR with the fix is merged. This should be released in Moya 13 in the near future! You can try it out using `develop` branch, but please don't use the code from develop on production yet. Thanks!
